DM-20915: Add str(Formatter) and force Formatter to require an arg for constructor

timj opened the pull request:
There is no longer any gain in allowing an optional FileDescriptor argument so remove that option.

timj committed:
Require a Formatter takes a parameter

Update tests to match, switching some of them to using
getFormatterClass so that the parameter is irrelevant.

timj committed:
Add str(Formatter)

TallJimbo commented on python/lsst/daf/butler/core/formatter.py:
```diff
@@ -50,11 +50,14 @@ class Formatter(metaclass=ABCMeta):
 
+    def __str__(self):
+        return f"{self.name()}@{self.fileDescriptor.location.path}"
+
```
Seems like it'd be straightforward to add the ideal kind of `__repr__`, too.

timj replied:
Yes, it was straightforward here but required I added repr() to FileDescriptor and Location and also that I cleaned up repr() for StorageClass to make it less chatty.  You now get a repr that looks something like:
```
lsst.daf.butler.formatters.jsonFormatter.JsonFormatter(FileDescriptor(Location('/Users/timj/work/lsstsw3/build/daf_butler/a', 'B'), StorageClass('xxx'), readStorageClass=StorageClass('yyy')))
```
@TallJimbo you might want to take a look at all the new code now.

TallJimbo replied:
Looks good.  I didn't know about `{!r}`, and I'm glad that I now do.

timj committed:
Add repr for Location and FileDescriptor

timj committed:
Change how getFullTypeName works with builtins

Having it return "builtins.str" serves no useful purpose
and it's much easier to return "str" and "dict".

timj committed:
Change StorageClass constructor so that it extracts name early

Previously the python type to name conversion was deferred until
the python type was needed.  This is unnecessary and causes
problems with repr() implementations. Now always extract the full
name using getFullTypeName.

timj committed:
Rewrite StorageClass.__repr__

Make it more compact by removing entities that are defaulted.

timj committed:
Add Formatter.__repr__

timj committed:
Update Instrument.getRawFormatter docstring

Now consistent in mentioning it returns a Formatter class not
Formatter instance.
